### ISSUE TITLE
Using QueueTaskDispatcher to block jobs using a locked resource before deploying to node 

### ISSUE THREAD

**Issue by maljaberi** [jenkinsci/lockable-resources-plugin#93]
Hi,

I don't know if I'm writing this in the correct place, but I have been wondering about how to solve the following problem:

Given:
A declarative Jenkinsfile 'A' that references label 'B':
`Pipeline { agent { label: 'B'} ... }`

One of the stages in the Pipeline takes Lock 'L'.

When:

- There exists N jobs using Jenkinsfile 'A'
- K nodes with labels 'B'

Then:
If any job of the set N takes the lock 'L', I want the rest  N-1 to be blocked with CauseOfBlockage set to BecauseLabelIsBusy.

So only 1 node should be running and the rest K-1 should be free for other jobs of another set to execute on.

Note: This is needed to lock resource to different hardware setups, in reality the lock string is a parameter passed by the job, some jobs that don't need any resource and will pass BUILD_TAG as a unique lock string.
This way we should be able to have higher utilization of the nodes and avoid having dedicated nodes for specially constrained jobs.

I have been looking around and I found two plugins that if combined in someway might give a good solution:
- Lockable plugin (to manage the locking mechanism)
- Build-blocker-plugin (to hook to QueueTaskDispatcher extension point)

Am I looking in the right direction? or can this be totally done by the Lockable plugin?

Thank you in advance for your feedback!

**Comment by dcendents** [comment]
@maljaberi I don't use declarative pipelines, I use full groovy pipelines but this is something you can do with only this plugin.

Just make sure you don't take a node until you have acquired the lock. Using full pipeline this would mean:

```groovy
lock(label: 'L') {
    node('B') {
        ...
    }
}
```

**Comment by maljaberi** [comment]
@dcendents  This does not block other jobs from allocating other executors with label 'B'. For my case, each executor runs on a different node.

I think I need something to lock on the label B itself, in a way to block jobs from being scheduled. 

Btw, what is the difference between declarative and fully groovy pipelines?

**Comment by dcendents** [comment]
@maljaberi Then I don't understand your use case. I though you said you wanted the rest (K-1) to be free to run other jobs.

In my opinion node labels should reflect capabilities to restrict where a job can be executed successfully. The lock plugin should be used to make sure you don't use more resources than are available (licenses, physical hardware, etc). Sometimes it can be hard to mix both so maybe make sure you use the same labels on both.

Job A requires resources with label B, if it acquires it it will require to run on a node with label B. If no node is available it will be in the queue until a node is free, during that time no one else will be able to lock resource B but other jobs already in the queue might get to run on node B before A.

If you want to reserve node B for jobs that acquire lock on resource B, change the node usage to run only jobs that match the label.

Finally the right name for fully groovy pipelines is "Scripted Pipelines". You can see the difference between the two here: https://jenkins.io/doc/book/pipeline/syntax/

**Comment by maljaberi** [comment]
> Then I don't understand your use case. I though you said you wanted the rest (K-1) to be free to run other jobs.

yes, I want them to run other jobs not from the N jobs that require the lock to a hardware resource

> In my opinion node labels should reflect capabilities to restrict where a job can be executed successfully. The lock plugin should be used to make sure you don't use more resources than are available

We don't disagree here :)

Maybe an example of the declarative pipeline will be more clear to discuss about:
`pipeline {
  agent { label: 'B' }
  stages {
      stage(1) { ... }
      stage(2) { 
         lock ('L') {...}
      }
      stage(3) {...}
   }
}`

Declarative pipeline requires me to define an executor to run in for the entire pipeline, so if I use your segguestion with `lock ('L') { node('B') { ... } }` in stage 2 it will only allocate another executor, so now I'm allocating twice as many.

Now in declarative, I cannot enclose the whole pipeline with a lock.

Nor can I enclose the 'stages' section with a lock. It will not help either because the job has already been scheduled in the start of the pipeline.

Conceptually I want to be able to add an option to the agent section telling it to lock the resource before scheduling the job, something like `agent { label: 'B' lock:'L' }

It looks like a limitation in the declarative pipeline?

**Comment by dcendents** [comment]
Yes I believe you will need to switch to scripted pipelines if you want to lock first and then take an executor (node). The advantage of the scripted pipelines is if your jobs are all similar, you can define a single script and add conditional blocks, that's what I do (single script, over 100 jobs, mostly maven, a few ants, a few don't have unit tests, etc)

Example:
```groovy
// Job specific configuration
def hasUnitTests = true
def hasIntegrationTests = true
def gitUrl = 'http://...'
def buildCommand = 'mvn clean install'
etc

// Global script
node {
  stage("compile") {
    // git clone
    // sh buildCommand
    if (hasUnitTests) {
      junit ...
      jacoco ...
    }
  }
}

if (hasIntegrationTests) {
  lock('wildfly') {
    node {
      stage("integration tests") {
        // git clone (might not get the previous workspace)
        ...
      }
    }
  }
}

...
```
the lock plugin guarantees the execution will wait there until the lock is acquired (single job if locking by name or multiple jobs if locking by label). The nodes are free to process all the `node` blocks from your jobs as the pipelines get executed.

In my example above, the jobs will always compile and execute unit tests, but only 1 job at a time will be able to execute its integration tests. Jobs without integration tests will never be blocked.

**Comment by maljaberi** [comment]
Thank you for sharing your method.

I use same pipeline too, it's defined as a global variable that then is imported as a library in a Jenkins file that will call the pipeline variable with different parameters for each job, one of which is the lock string.

If I would write my own plugin for declarative pipeline to lock before taking an executor, I suppose I would call this plugin with lock operation somehow. Is this achievable in someway? If yes, can you guide me where to start?

**Comment by jimklimov** [comment]
I think the idea discussion above misses one crucial point: if you want the node which fulfills "node label B" to only be used by one build at a time to avoid conflicts, you can define its node properties as having one executor (it may run parallel make or whatever as children of that executor thread). That way the example pipeline above would first lock a resource (say printer), then get a node by label (say, has printer drivers), and while it uses that node nobody else can use it, neither by name nor by label match - the only executor there is busy. There may be many other similar nodes with same label that would still be available, one executor each...
